File: Data Structures & Algorithms/two-integer-sum/submission-0.js
class Solution {
    /**
     * @param {number[]} nums
     * @param {number} target
     * @return {number[]}
     */
    naiveApproach(nums, target) {

    }
    optimizedApproach(nums, target) {
        let pairs = new Map();
        for (let i = 0; i < nums.length; i++) {
            let diff = target - nums[i];
            if (pairs.has(diff)) {
                return [i, pairs.get(diff)];
            }
            pairs.set(nums[i], i);
        }
        return []
    }
    twoSum(nums, target) {
        return this.optimizedApproach.apply({}, [nums, target]);
    }
}

File: Data Structures & Algorithms/target-sum/submission-0.java
public class Solution {
    public int findTargetSumWays(int[] nums, int target) {
        return backtrack(0, 0, nums, target);
    }

    private int backtrack(int i, int total, int[] nums, int target) {
        if (i == nums.length) {
            return total == target ? 1 : 0;
        }
        return backtrack(i + 1, total + nums[i], nums, target) +
               backtrack(i + 1, total - nums[i], nums, target);
    }
}